### PR TITLE
Add a check for trailing slashes in endpoint URL.

### PR DESCRIPTION
**Pull request by ShalkiWenushika**
### Issues

Fixes https://github.com/wso2/product-microgateway/issues/2168

### Automation tests
 - Unit tests added: Yes/No
 - Integration tests added: Yes/No

### Tested environments

Not Tested

---
#### Maintainers: Check before merge
- [x] Assigned 'Type' label
- [x] Assigned the project
- [x] Validated respective github issues
- [x] Assigned milestone to the github issue(s)

### CODE REVIEW

File: adapter/internal/oasparser/envoyconf/routes_with_clusters.go [hunk 1]
```diff
@@ -162,6 +162,10 @@ func CreateRoutesWithClusters(mgwSwagger model.MgwSwagger, upstreamCerts []byte,
 			logger.LoggerOasparser.Warnf("Production environment endpoints are not available for the resource %v:%v-%v",
 				apiTitle, apiVersion, resource.GetPath())
 		}
+		// Check whether the endpoint contains a trailing slash.
+		if string(endpointBasepath) == "/" {
```
Comment (VirajSalaka) [inline adapter/internal/oasparser/envoyconf/routes_with_clusters.go:166]: Should not we generalize such that we check for the trailing `/`

Comment (ShalkiWenushika) [reply adapter/internal/oasparser/envoyconf/routes_with_clusters.go]: Fixed via https://github.com/wso2/product-microgateway/pull/2167/commits/d3449286f728bf04395dd35a9c6701ce19c3a36a

File: adapter/internal/oasparser/envoyconf/routes_with_clusters.go [hunk 2]
```diff
@@ -162,6 +162,13 @@ func CreateRoutesWithClusters(mgwSwagger model.MgwSwagger, upstreamCerts []byte,
 			logger.LoggerOasparser.Warnf("Production environment endpoints are not available for the resource %v:%v-%v",
 				apiTitle, apiVersion, resource.GetPath())
 		}
+		// Check whether the endpoint basepath contains a trailing slash.
+		if endpointBasepath != "" {
```
Comment (VirajSalaka) [inline adapter/internal/oasparser/envoyconf/routes_with_clusters.go:166]: you could use hasSuffix method here.
https://pkg.go.dev/strings

Comment (VirajSalaka) [reply adapter/internal/oasparser/envoyconf/routes_with_clusters.go:166]: Or rather TrimSuffix straightaway. Then that would be a single line code.
https://pkg.go.dev/strings#TrimSuffix

Comment (ShalkiWenushika) [reply adapter/internal/oasparser/envoyconf/routes_with_clusters.go:166]: Resolved via https://github.com/wso2/product-microgateway/pull/2167/commits/b764eee8b572ae14f67dff9a41c136dc1c1807e9

**Review by menakaj — APPROVED** [review]

File: integration/test-integration/src/test/java/org/wso2/choreo/connect/tests/testcases/standalone/endpoints/EndpointsWithTrailingSlashTestCase.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+import org.wso2.choreo.connect.tests.util.ApictlUtils;
+import org.wso2.choreo.connect.tests.util.HttpResponse;
+import org.wso2.choreo.connect.tests.util.HttpsClientRequest;
+import org.wso2.choreo.connect.tests.util.Utils;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+
+public class EndpointsWithTrailingSlashTestCase {
```
Comment (Rajith90) [inline integration/test-integration/src/test/java/org/wso2/choreo/connect/tests/testcases/standalone/endpoints/EndpointsWithTrailingSlashTestCase.java:36]: Can we move this test case to the APIM test suite

Comment (ShalkiWenushika) [reply integration/test-integration/src/test/java/org/wso2/choreo/connect/tests/testcases/standalone/endpoints/EndpointsWithTrailingSlashTestCase.java:36]: In APIM 4.0.0 this issue is not reproducing. That's why I implemented this test case in the standalone test suite.

Comment (ShalkiWenushika) [reply integration/test-integration/src/test/java/org/wso2/choreo/connect/tests/testcases/standalone/endpoints/EndpointsWithTrailingSlashTestCase.java]: Issue was reproduced in the latest APIM 4.0.0 pack. Fixed via https://github.com/wso2/product-microgateway/pull/2167/commits/9a417cff52b8ac407fb20039e5c40c8e5639489f

**Review by Rajith90 — APPROVED** [review]

**Review by Rajith90 — APPROVED** [review]

**Comment by choreo-cicd** [comment]
[succeeded] dev-deployment : 20210726.19

**Comment by choreo-cicd** [comment]
[succeeded] stage-deployment : 20210727.1

**Comment by choreo-cicd** [comment]
[succeeded] prod-deployment : 20210727.1

**Pull request merged**